File: source/duelink-ai-stdlib/python/duelink_ai/engine.py
import os
import json
import aiohttp


class Engine:
    def __init__(self, key, jsonPath, duelink):
        self.apiKey = key
        self.lastPin = 1
        self.duelink = duelink

        if os.path.exists(jsonPath):
            self.jsonFiles = [
                os.path.join(jsonPath, f)
                for f in os.listdir(jsonPath)
                if f.endswith(".json")
            ]
        else:
            self.jsonFiles = []

        self.knownKeywords = [
            "pin","led","blink","high","low",
            "analog","voltage",
            "beep","sound",
            "frequency","pwm",
            "info","version",
            "file","fs","read","write",
            "draw","display","text",
            "temperature","humidity",
            "servo","motor",
            "distance",
            "interrupt",
            "convert","scale","hex",
            "downlink","cmd",
            "pid","product id","firmware",
            "reset","erase"
        ]

        self.toolDictionaryMap = {
            "statusled": ["led", "blink", "flash", "pin", "high", "low", "on", "off"],
            "digital": ["digital", "pin", "high", "low", "on", "off"],
            "analog": ["analog", "volt", "voltage", "adc", "vcc"],
            "sound": ["beep", "sound", "buzzer", "tone", "play", "hz", "khz", "mhz"],
            "frequency": ["frequency", "pwm", "duty", "cyle", "dutycyle", "hz", "khz", "mhz"],
            "system": ["version", "info", "reset", "device", "pid", "product id", "firmware"],
            "i2c": ["i2c", "two wire", "bus"],
            "spi": ["spi"],
            "uart": ["uart", "serial", "tx", "rx"],
            "deviceaddressing": ["sel", "address", "device address"],
            "filesystem": ["file", "fs", "read file", "write file", "directory"],
            "graphics": ["graphics", "draw", "display", "pixel", "text", "circle", "screen", "show", "clear", "fill", "line", "color", "image"],
            "math": ["sin", "cos", "tan", "sqrt", "random"],
            "converter": ["convert", "hex", "format", "scale", "base64"],
            "interrupt": ["interrupt", "irq", "trigger", "callback"],
            "downlinkcontrol": ["downlink"],
            "button": ["button", "press"],
            "distance": ["distance", "ultrasonic", "trigger", "echo"],
            "servomotor": ["servo", "motor angle", "degree"],
            "temperature": ["temperature", "temp", "dht"],
            "humidity": ["humidity"],
            "print": ["print", "println"]
        }

        self.strictRules = (
            "STRICT RULES:"
            "\n- ALWAYS call a function"
            "\n- NEVER return text"
            "\n- Use ONLY provided functions"
            "\n- DO NOT invent names"
            "\n- Follow parameters EXACTLY"
            "\n- Arrays must be [ ... ]"
            "\n- 1 second = 1000"
            "\n- on=1, off=0"
            "\n- Use milliseconds (1 second = 1000)"
            "\n- If user says 'on' → value=1"
            "\n- If user says 'off' → value=0"
            "\n- Follow parameter meaning strictly"
            "\n- Use enums when available"
        )

        self.Verbose = True

    # =========================
    async def Run(self, prompt):
        prompt = self.Normalize(prompt)

        topics = self.GetSupportedTopics(self.jsonFiles)

        if (not self.IsHelpRequest(prompt)) and self.IsUnknownRequest(prompt, self.jsonFiles):
            return [f"This request is not supported. \nEnhance the JSON files, KnownKeywords, and ToolDictionaryMap properties to improve intelligence and accuracy"]

        toolsFile = self.SelectTool(prompt, self.jsonFiles)
        tools = open(toolsFile, "r", encoding="utf-8").read()

        if self.IsHelpRequest(prompt):
            help_text = await self.CallOpenAIHelp(prompt, tools)
            return [help_text]

        response = await self.CallOpenAI(prompt, tools)
        cmds = self.ParseTools(response)

        result = []
        for cmd in cmds:
            result.append(self.Execute(cmd))

        return result

    # =========================
    def Normalize(self, input):
        input = input.lower()
        input = input.replace("one second", "1000 ms")
        input = input.replace("1 second", "1000 ms")
        input = input.replace("half second", "500 ms")
        input = input.replace("turn on", "set high")
        input = input.replace("turn off", "set low")
        return input

    def IsHelpRequest(self, prompt):
        prompt = prompt.lower()
        return ("help" in prompt or
                "what is" in prompt or
                "explain" in prompt or
                "how to use" in prompt)

    def GetSupportedTopics(self, files):
        topics = list(set([
            os.path.splitext(os.path.basename(f))[0].lower()
            for f in files
        ]))
        return ", ".join(topics)

    def IsUnknownRequest(self, prompt, files):
        prompt = prompt.lower()

        for f in files:
            name = os.path.splitext(os.path.basename(f))[0].lower()
            if name in prompt:
                return False

        if any(k in prompt for k in self.knownKeywords):
            return False

        return True

    # =========================
    def BuildHelpPrompt(self, prompt, toolJson):
        return f"""
User question: {prompt}

Here is the API definition:
{toolJson}

Explain this function for a beginner.

FORMAT STRICTLY:

1. Title
2. Simple explanation (very easy)
3. Parameters (bullet points)

4. Command (ready to run)
- Provide ONE real command line (like: statled(1000,1000,0))

5. Explanation of the command
- Explain what each value does in that example

6. What happens in real life

7. Suggest 1 thing to try next

Rules:
- Keep it simple
- Time must be milliseconds (1000 = 1 second)
- If duration=0 or count=0 means forever, explain it
- Command must be VALID and realistic
- DO NOT call function (this is explanation mode)
"""

    async def CallOpenAIHelp(self, prompt, toolsJson):
        headers = {
            "Authorization": f"Bearer {self.apiKey}",
            "Content-Type": "application/json"
        }

        request = {
            "model": "gpt-4o-mini",
            "messages": [
                {"role": "system", "content": "You are a friendly DUELink teacher helping beginners understand hardware APIs."},
                {"role": "user", "content": self.BuildHelpPrompt(prompt, toolsJson)}
            ]
        }

        async with aiohttp.ClientSession() as session:
            async with session.post("https://api.openai.com/v1/chat/completions", headers=headers, json=request) as res:
                data = await res.json()
                return data["choices"][0]["message"]["content"]

    # =========================
    def SelectTool(self, prompt, files):
        prompt = prompt.lower()
        candidates = []

        for k, v in self.toolDictionaryMap.items():
            if any(x in prompt for x in v):
                candidates.append(k)

        if len(candidates) == 0:
            return self.FindFile(files, "digital") or files[0]

        if len(candidates) == 1:
            return self.FindFile(files, candidates[0])

        scores = {}

        for c in candidates:
            my = self.toolDictionaryMap[c]
            others = [x for k in candidates if k != c for x in self.toolDictionaryMap[k]]

            score = sum(1 for k in my if k in prompt and k not in others)
            scores[c] = score

        maxScore = max(scores.values())
        best = [k for k, v in scores.items() if v == maxScore]

        if len(best) == 1:
            return self.FindFile(files, best[0])

        bestMatch = None
        bestIndex = 999999

        for c in best:
            for k in self.toolDictionaryMap[c]:
                idx = prompt.find(k)
                if idx >= 0 and idx < bestIndex:
                    bestIndex = idx
                    bestMatch = c

        if bestMatch:
            return self.FindFile(files, bestMatch)

        return self.FindFile(files, "digital") or files[0]

    def FindFile(self, files, keyword):
        keyword = keyword.lower()

        for f in files:
            name = os.path.splitext(os.path.basename(f))[0].lower()
            if name == keyword:
                return f

        for f in files:
            name = os.path.splitext(os.path.basename(f))[0].lower()
            if keyword in name:
                return f

        return None

    # =========================
    async def CallOpenAI(self, prompt, toolsJson):
        headers = {
            "Authorization": f"Bearer {self.apiKey}",
            "Content-Type": "application/json"
        }

        tools = json.loads(toolsJson)["tools"]

        request = {
            "model": "gpt-4o-mini",
            "messages": [
                {"role": "system", "content": "You are a DUELink hardware assistant.\n\nConvert user request into function calls.\n\n" + self.strictRules},
                {"role": "user", "content": prompt}
            ],
            "tools": tools,
            "tool_choice": "required"
        }

        async with aiohttp.ClientSession() as session:
            async with session.post("https://api.openai.com/v1/chat/completions", headers=headers, json=request) as res:
                return await res.text()

    # =========================
    def ParseTools(self, json_str):
        doc = json.loads(json_str)

        if "error" in doc:
            print("OpenAI Error:", doc["error"]["message"])
            return []

        if "choices" not in doc:
            print("Invalid response:", json_str)
            return []

        message = doc["choices"][0].get("message", {})

        if "tool_calls" not in message:
            print("No tool call:", message.get("content"))
            return []

        result = []

        for t in message["tool_calls"]:
            name = t["function"]["name"]
            args = json.loads(t["function"]["arguments"])
            result.append((name, args))

        return result

    # =========================
    def Execute(self, cmd):
        name, args = cmd

        command = name + "("

        if isinstance(args, dict):
            values = []

            for k in args:
                val = self.FormatValue(args[k])

                if k == "pin":
                    try:
                        self.lastPin = int(val)
                    except:
                        pass

                values.append(val)

            command += ",".join(values)

        command += ")"

        self.Log("SEND: " + command)

        if self.duelink:
            ret = self.duelink.Engine.ExecuteCommandRaw(command)

            if ret:
                self.Log("RETURN: " + ret)

            return ret

        return ""

    # =========================
    def FormatValue(self, val):
        if isinstance(val, (int, float)):
            return str(val)
        if isinstance(val, str):
            return f'"{val}"'
        if isinstance(val, list):
            return "[" + ",".join(str(x) for x in val) + "]"
        if isinstance(val, bool):
            return "1" if val else "0"
        return str(val)

    # =========================
    def Log(self, log):
        if self.Verbose:
            print(log)

File: source/duelink-ai-stdlib/dotnet/GHIElectronics.DUELink.AI/Engine.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Reflection.Metadata.Ecma335;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace GHIElectronics.DUELink.AI
{
    internal class Engine
    {
        private string apiKey;
        private int lastPin = 1;
        private string[] jsonFiles;
        private DUELinkController duelink;

        public Engine(string key, string jsonPath, DUELinkController duelink)
        {
            apiKey = key;

            var dir = jsonPath;
            if (Directory.Exists(dir))
                jsonFiles = Directory.GetFiles(dir, "*.json");
            else
                jsonFiles = Array.Empty<string>();

            this.duelink = duelink; 


        }

        public async Task<string[]> Run(string prompt)
        {
            prompt = Normalize(prompt);

            var topics = GetSupportedTopics(jsonFiles);

            if (!IsHelpRequest(prompt) && IsUnknownRequest(prompt, jsonFiles))
            {
                return new[] { $"This request is not supported. \nEnhance the JSON files, KnownKeywords, and ToolDictionaryMap properties to improve intelligence and accuracy" };
            }

            var toolsFile = SelectTool(prompt, jsonFiles);
            var tools = File.ReadAllText(toolsFile);

            if (IsHelpRequest(prompt))
            {
                var help = await CallOpenAIHelp(prompt, tools);
                return new[] { help };
            }

            var response = await CallOpenAI(prompt, tools);
            var cmds = ParseTools(response);

            var list = new List<string>();

            foreach (var cmd in cmds)
            {
                list.Add(Execute(cmd));
            }

            return list.ToArray();
        }

        // =========================
        // NORMALIZE INPUT
        // =========================
        string Normalize(string input)
        {
            input = input.ToLower();

            // time conversion
            input = input.Replace("one second", "1000 ms");
            input = input.Replace("1 second", "1000 ms");
            input = input.Replace("half second", "500 ms");

            // intent mapping
            input = input.Replace("turn on", "set high");
            input = input.Replace("turn off", "set low");

            return input;
        }

        bool IsHelpRequest(string prompt)
        {
            prompt = prompt.ToLower();

            return prompt.Contains("help") ||
                   prompt.Contains("what is") ||
                   prompt.Contains("explain") ||
                   prompt.Contains("how to use");
        }



        string GetSupportedTopics(string[] files)
        {
            var topics = files
                .Select(f => Path.GetFileNameWithoutExtension(f).ToLower())
                .Distinct();

            return string.Join(", ", topics);
        }

        string[] knownKeywords = {
            "pin","led","blink","high","low",
            "analog","voltage",
            "beep","sound",
            "frequency","pwm",
            "info","version",
            "file","fs","read","write",
            "draw","display","text",
            "temperature","humidity",
            "servo","motor",
            "distance",
            "interrupt",
            "convert","scale","hex",
            "downlink","cmd",
            "pid", "product id","firmware",
            "reset", "erase"
        };

        public string[] KnownKeywords {
            get { return knownKeywords; }
            set { knownKeywords = value; }
        
        }

        bool IsUnknownRequest(string prompt, string[] files)
        {
            prompt = prompt.ToLower();

            foreach (var f in files)
            {
                var name = Path.GetFileNameWithoutExtension(f).ToLower();

                if (prompt.Contains(name))
                    return false;
            }

            if (knownKeywords.Any(k => prompt.Contains(k)))
                return false;

            return true;
        }        

        string BuildHelpPrompt(string prompt, string toolJson)
        {
            return $@"
User question: {prompt}

Here is the API definition:
{toolJson}

Explain this function for a beginner.

FORMAT STRICTLY:

1. Title
2. Simple explanation (very easy)
3. Parameters (bullet points)

4. Command (ready to run)
- Provide ONE real command line (like: statled(1000,1000,0))

5. Explanation of the command
- Explain what each value does in that example

6. What happens in real life

7. Suggest 1 thing to try next

Rules:
- Keep it simple
- Time must be milliseconds (1000 = 1 second)
- If duration=0 or count=0 means forever, explain it
- Command must be VALID and realistic
- DO NOT call function (this is explanation mode)
";
        }

        async Task<string> CallOpenAIHelp(string prompt, string toolsJson)
        {
            var http = new HttpClient();
            http.DefaultRequestHeaders.Add("Authorization", $"Bearer {apiKey}");

            var request = new
            {
                model = "gpt-4o-mini",
                messages = new object[]
                {
            new {
                role = "system",
                content = "You are a friendly DUELink teacher helping beginners understand hardware APIs."
            },
            new {
                role = "user",
                content = BuildHelpPrompt(prompt, toolsJson)
            }
                }
            };

            var content = new StringContent(JsonSerializer.Serialize(request), Encoding.UTF8, "application/json");

            var res = await http.PostAsync("https://api.openai.com/v1/chat/completions", content);
            var json = await res.Content.ReadAsStringAsync();

            using var doc = JsonDocument.Parse(json);

            return doc.RootElement
                .GetProperty("choices")[0]
                .GetProperty("message")
                .GetProperty("content")
                .GetString();
        }

        // =========================
        // SELECT TOOL
        // =========================

        Dictionary<string, string[]> toolDictionaryMap = new Dictionary<string, string[]>
        {
            ["statusled"] = new[] { "led", "blink", "flash", "pin", "high", "low", "on", "off" },
            ["digital"] = new[] { "digital", "pin", "high", "low", "on", "off" },
            ["analog"] = new[] { "analog", "volt", "voltage", "adc", "vcc" },
            ["sound"] = new[] { "beep", "sound", "buzzer", "tone", "play", "hz", "khz", "mhz" },
            ["frequency"] = new[] { "frequency", "pwm", "duty", "cyle", "dutycyle", "hz", "khz", "mhz" },
            ["system"] = new[] { "version", "info", "reset", "device", "pid", "product id", "firmware" },
            ["i2c"] = new[] { "i2c", "two wire", "bus" },
            ["spi"] = new[] { "spi" },
            ["uart"] = new[] { "uart", "serial", "tx", "rx" },
            ["deviceaddressing"] = new[] { "sel", "address", "device address" },
            ["filesystem"] = new[] { "file", "fs", "read file", "write file", "directory" },
            ["graphics"] = new[] { "graphics", "draw", "display", "pixel", "text", "circle", "screen", "show", "clear", "fill", "line", "color", "image" },
            ["math"] = new[] { "sin", "cos", "tan", "sqrt", "random" },
            ["converter"] = new[] { "convert", "hex", "format", "scale", "base64" },
            ["interrupt"] = new[] { "interrupt", "irq", "trigger", "callback" },
            ["downlinkcontrol"] = new[] { "downlink" },
            ["button"] = new[] { "button", "press" },

            ["distance"] = new[] { "distance", "ultrasonic", "trigger", "echo" },

            ["servomotor"] = new[] { "servo", "motor angle", "degree" },
            ["temperature"] = new[] { "temperature", "temp", "dht" },
            ["humidity"] = new[] { "humidity" },
            ["print"] = new[] { "print", "println" }
        };

        public Dictionary<string, string[]> ToolDictionaryMap {  
            get => toolDictionaryMap;
            set => toolDictionaryMap = value; 
        }
        string SelectTool(string prompt, string[] files)
        {
            prompt = prompt.ToLower();

            var map = toolDictionaryMap;

            // ===== STEP 1: find candidates =====
            var candidates = new List<string>();

            foreach (var kv in map)
            {
                if (kv.Value.Any(k => prompt.Contains(k)))
                    candidates.Add(kv.Key);
            }

            if (candidates.Count == 0)
                return FindFile(files, "digital") ?? files[0];

            if (candidates.Count == 1)
                return FindFile(files, candidates[0]);

            // ===== STEP 2: unique keyword scoring =====
            var scores = new Dictionary<string, int>();

            foreach (var c in candidates)
            {
                var myKeywords = map[c];
                var others = candidates.Where(x => x != c).SelectMany(x => map[x]).ToList();

                int uniqueScore = myKeywords
                    .Where(k => prompt.Contains(k) && !others.Contains(k))
                    .Count();

                scores[c] = uniqueScore;
            }

            int maxScore = scores.Values.Max();

            var best = scores.Where(x => x.Value == maxScore).Select(x => x.Key).ToList();

            if (best.Count == 1)
                return FindFile(files, best[0]);

            // ===== STEP 3: earliest keyword wins =====
            string bestMatch = null;
            int bestIndex = int.MaxValue;

            foreach (var c in best)
            {
                foreach (var k in map[c])
                {
                    int idx = prompt.IndexOf(k);
                    if (idx >= 0 && idx < bestIndex)
                    {
                        bestIndex = idx;
                        bestMatch = c;
                    }
                }
            }

            if (bestMatch != null)
                return FindFile(files, bestMatch);

            return FindFile(files, "digital") ?? files[0];
        }

        string FindFile(string[] files, string keyword)
        {
            keyword = keyword.ToLower();

            // 1Exact match FIRST
            var exact = files.FirstOrDefault(f =>
                Path.GetFileNameWithoutExtension(f)
                    .Equals(keyword, StringComparison.OrdinalIgnoreCase));

            if (exact != null)
                return exact;

            // Fallback: contains
            return files.FirstOrDefault(f =>
                Path.GetFileNameWithoutExtension(f)
                    .ToLower()
                    .Contains(keyword));
        }

        // =========================
        // CALL OPENAI
        // =========================

        string strictRules = "STRICT RULES:" +
            "\n- ALWAYS call a function" +
            "\n- NEVER return text" +
            "\n- Use ONLY provided functions" +
            "\n- DO NOT invent names" +
            "\n- Follow parameters EXACTLY" +
            "\n- Arrays must be [ ... ]" +
            "\n- 1 second = 1000" +
            "\n- on=1, off=0" +
            "\n- Use milliseconds (1 second = 1000)" +
            "\n- If user says 'on' → value=1" +
            "\n- If user says 'off' → value=0" +
            "\n- Follow parameter meaning strictly" +
            "\n- Use enums when available";
        public string StrictRules {  
            get { return strictRules; } 
            set { strictRules = value; }
        
        }
        async Task<string> CallOpenAI(string prompt, string toolsJson)
        {
            var http = new HttpClient();
            http.DefaultRequestHeaders.Add("Authorization", $"Bearer {apiKey}");

            using var doc = JsonDocument.Parse(toolsJson);
            var _tools = doc.RootElement.GetProperty("tools").EnumerateArray().ToArray();

            var request = new
            {
                model = "gpt-4o-mini",
                messages = new object[]
                {
                new {
                    role = "system",
                    content = @"You are a DUELink hardware assistant.

Convert user request into function calls.

" + strictRules
                },
                new { role = "user", content = prompt }
                },
                tools = _tools,
                tool_choice = "required"
            };

            var content = new StringContent(JsonSerializer.Serialize(request), Encoding.UTF8, "application/json");

            var res = await http.PostAsync("https://api.openai.com/v1/chat/completions", content);
            return await res.Content.ReadAsStringAsync();
        }

        // =========================
        // PARSE MULTI TOOL
        // =========================
        (string name, JsonElement args)[] ParseTools(string json)
        {
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;

            //  CHECK ERROR FIRST
            if (root.TryGetProperty("error", out var err))
            {
                Console.WriteLine(" OpenAI Error:");
                Console.WriteLine(err.GetProperty("message").GetString());
                return Array.Empty<(string, JsonElement)>();
            }

            //  CHECK choices
            if (!root.TryGetProperty("choices", out var choices))
            {
                Console.WriteLine(" Invalid response: no choices");
                Console.WriteLine(json);
                return Array.Empty<(string, JsonElement)>();
            }

            var choice0 = choices[0];

            if (!choice0.TryGetProperty("message", out var message))
            {
                Console.WriteLine(" Invalid response: no message");
                Console.WriteLine(json);
                return Array.Empty<(string, JsonElement)>();
            }

            //  CHECK tool_calls
            if (!message.TryGetProperty("tool_calls", out var toolCalls))
            {
                Console.WriteLine(" No tool call. AI said:");
                if (message.TryGetProperty("content", out var content))
                    Console.WriteLine(content.GetString());

                return Array.Empty<(string, JsonElement)>();
            }

            var list = new List<(string, JsonElement)>();

            foreach (var tool in toolCalls.EnumerateArray())
            {
                var name = tool.GetProperty("function").GetProperty("name").GetString();
                var argsString = tool.GetProperty("function").GetProperty("arguments").GetString();
                var args = JsonDocument.Parse(argsString).RootElement;

                list.Add((name, args));
            }

            return list.ToArray();
        }

        // =========================
        // EXECUTE COMMAND
        // =========================
        string Execute((string name, JsonElement args) cmd)
        {
            if (cmd.name == "print" || cmd.name == "println")
            {
                if (cmd.args.TryGetProperty("values", out var values))
                {
                    // if only 1 element
                    if (values.GetArrayLength() == 1)
                    {
                        var v = values[0];

                        // number → convert to string
                        if (v.ValueKind == JsonValueKind.Number)
                        {                            
                            Log($"SEND: {cmd.name}(\"{v}\")");
                        }

                        // string → keep as is
                        if (v.ValueKind == JsonValueKind.String)
                        {

                            Log($"SEND: {cmd.name}(\"{v.GetString()}\")");
                        }
                    }

                    // multiple elements → array
                    var arr = values.EnumerateArray().Select(x => x.ToString());

        
                    Log($"SEND: {cmd.name}([{string.Join(",", arr)}])");
                }                
            }

            string command = cmd.name + "(";

            if (cmd.args.ValueKind == JsonValueKind.Object)
            {
                var values = new List<string>();

                foreach (var p in cmd.args.EnumerateObject())
                {
                    string val = FormatValue(p.Value);

                    if (p.Name == "pin" && int.TryParse(val, out int pin))
                        lastPin = pin;

                    values.Add(val);
                }

                command += string.Join(",", values);
            }

            command += ")";

            //Log("SEND: " + command);
            if (duelink == null)
            {
                Log("SEND: " + command);
            }
            else
            {
                Log("SEND: " + command);
                var ret = duelink.Engine.ExecuteCommandRaw(command);

                if ( ret != null && ret.Length > 0)
                {
                    Log("RETURN: " + ret);
                }
                
                
                return ret;
            }

            // TODO: serial.WriteCommand(command);
            return "";
        }

        // =========================
        // FIX VALUE
        // =========================
        string FixValue(string v)
        {
            v = v.ToLower();

            if (v == "high" || v == "on") return "1";
            if (v == "low" || v == "off") return "0";

            return v;
        }

        string FormatValue(JsonElement val)
        {
            switch (val.ValueKind)
            {
                case JsonValueKind.Number:
                    return val.ToString();

                case JsonValueKind.String:
                    return $"\"{val.GetString()}\""; //  always quote

                case JsonValueKind.Array:
                    var items = val.EnumerateArray().Select(x => x.ToString());
                    return "[" + string.Join(",", items) + "]";

                case JsonValueKind.True:
                    return "1";

                case JsonValueKind.False:
                    return "0";

                default:
                    return val.ToString();
            }
        }

        public bool Verbose { get; set; } = true;
        void Log(string log) { 
            if (Verbose) Console.WriteLine(log);
            
        }    
    }
}
